plugins/wasi_nn/wasinn_ggml.cpp

```diff
@@ -396,8 +396,73 @@ void buildOutputEmbedding(std::string &Embedding, int32_t NEmbd,
                   NEmbd, fmt::join(Embeddings, Embeddings + NEmbd, ","sv));
 }
 
+static bool evaluateQwen2vlImageEmbed(
+    llama_context *CtxLlama, const struct llava_image_embed *ImageEmbed,
+    int NBatch, int *NPast, int *StPosId, struct clip_image_size *ImageSize) {
+  int NEmbd = llama_n_embd(llama_get_model(CtxLlama));
+  const int PatchSize = 14 * 2;
+  const int Ph =
+      ImageSize->height / PatchSize + (ImageSize->height % PatchSize > 0);
+  const int Pw =
+      ImageSize->width / PatchSize + (ImageSize->width % PatchSize > 0);
+  auto ImgTokens = ImageEmbed->n_image_pos;
+  std::vector<llama_pos> MRopePos;
+  MRopePos.resize(ImgTokens * 4);
+
+  for (int Y = 0; Y < Ph; Y++) {
+    for (int X = 0; X < Pw; X++) {
+      int I = Y * Pw + X;
+      MRopePos[I] = *StPosId;
+      MRopePos[I + ImgTokens] = *StPosId + Y;
+      MRopePos[I + ImgTokens * 2] = *StPosId + X;
+      MRopePos[I + ImgTokens * 3] = 0;
+    }
+  }
+  *StPosId += std::max(Pw, Ph);
+
+  int Processed = 0;
+  std::vector<llama_pos> BatchMRopePos;
+  BatchMRopePos.resize(ImgTokens * 4);
+
+  for (int I = 0; I < ImgTokens; I += NBatch) {
+    int NEval = ImgTokens - I;
+    if (NEval > NBatch) {
+      NEval = NBatch;
+    }
+
+    std::fill(BatchMRopePos.begin(), BatchMRopePos.end(), 0);
+    memcpy(BatchMRopePos.data(), &MRopePos[Processed],
+           NEval * sizeof(llama_pos));
+    memcpy(&BatchMRopePos[NEval * 1], &MRopePos[ImgTokens * 1 + Processed],
+           NEval * sizeof(llama_pos));
+    memcpy(&BatchMRopePos[NEval * 2], &MRopePos[ImgTokens * 2 + Processed],
+           NEval * sizeof(llama_pos));
+    memcpy(&BatchMRopePos[NEval * 3], &MRopePos[ImgTokens * 3 + Processed],
+           NEval * sizeof(llama_pos));
```

Comment:
Should we use `std::copy` or `std::copy_n` instead?

Reply:
Use `std::copy_n` instead.

Reply:
It seems like `memcpy` here

Reply:
My mistake. Code updated.